File: Sources/Onboarding/OnboardingFlowState.swift
import Foundation

struct OnboardingFlowState: Equatable {
    let pageCount: Int
    var selectedIndex: Int

    init(pageCount: Int, selectedIndex: Int = 0) {
        let normalizedPageCount = max(0, pageCount)

        self.pageCount = normalizedPageCount
        self.selectedIndex = min(max(0, selectedIndex), max(0, normalizedPageCount - 1))
    }

    var isLastPage: Bool {
        selectedIndex >= pageCount - 1
    }

    @discardableResult
    mutating func advance() -> Bool {
        guard !isLastPage else {
            return false
        }

        selectedIndex += 1
        return true
    }
}

File: onboarding-android/src/main/kotlin/com/vadimkomis/onboarding/OnboardingFlowState.kt
package com.vadimkomis.onboarding

internal class OnboardingFlowState(
    pageCount: Int,
    selectedIndex: Int = 0,
) {
    val pageCount: Int = pageCount.coerceAtLeast(0)

    var selectedIndex: Int = selectedIndex.coerceIn(
        minimumValue = 0,
        maximumValue = (this.pageCount - 1).coerceAtLeast(0),
    )
        private set

    val isLastPage: Boolean
        get() = selectedIndex >= pageCount - 1

    fun advance(): Boolean {
        if (isLastPage) {
            return false
        }

        selectedIndex += 1
        return true
    }
}
